Refactor periodic task

Jackie-Jiang opened the pull request:
Refactor periodic task to fix the following issues：
- PinotTaskManager.scheduleTasks() has no effect on non-leader controller, but returns result from previous run
  The reason for this is that, we store states for each run as member variables in the periodic task, which mixed the concept of task and run
- After calling stop() (lose leadership), when re-gaining leadership, no start() method is called to set up the environment
- Potential race condition

Replace init() with start() and call it whenever the leadership is gained
Move the basic control methods for periodic task (start(), run(), stop()) into BasePeriodicTask
Keep table level methods in ControllerPeriodicTask
For per-run states, add context generic type to pass them between processing multiple tables

jackjlli commented on pinot-core/src/main/java/org/apache/pinot/core/periodictask/BasePeriodicTask.java:
```diff
@@ -42,9 +61,125 @@ public long getInitialDelayInSeconds() {
+  }
+
+  /**
+   * Executes the task. This method should early terminate if {@link}
```
Please add the condition when it should early terminate.

Jackie-Jiang replied:
Good catch. Seems like where I took a break and ran for dinner haha

jackjlli commented on pinot-core/src/main/java/org/apache/pinot/core/periodictask/BasePeriodicTask.java:
```diff
@@ -42,9 +61,125 @@ public long getInitialDelayInSeconds() {
+  /**
+   * Can be override for extra task cleanups.
+   */
+  protected void stopTask() {
```
CleanupTask might be better?

Jackie-Jiang replied:
Done.

jackjlli commented on pinot-core/src/main/java/org/apache/pinot/core/periodictask/PeriodicTaskScheduler.java:
```diff
@@ -37,16 +37,14 @@
   /**
-   * Initialize the PeriodicTaskScheduler with list of PeriodicTasks
-   * @param periodicTasks
+   * Initialize the PeriodicTaskScheduler with list of PeriodicTasks.
```
Initializes

Jackie-Jiang replied:
Done.

jackjlli commented on pinot-core/src/main/java/org/apache/pinot/core/periodictask/BasePeriodicTask.java:
```diff
@@ -42,9 +61,125 @@ public long getInitialDelayInSeconds() {
   @Override
-  public String getTaskName() {
-    return _taskName;
+  public final synchronized void start() {
```
So the purpose of adding synchronized here is to avoid different periodic tasks running at the same time? Then is it the same as running multiple tasks in one thread?

Jackie-Jiang replied:
No, the purpose of this synchronized is to avoid calling start() and stop() at the same time. Basically another level of protection. (start() and stop() are called in scheduler, not in the task thread)

jackjlli commented on pinot-core/src/main/java/org/apache/pinot/core/periodictask/BasePeriodicTask.java:
```diff
@@ -42,9 +61,125 @@ public long getInitialDelayInSeconds() {
+      long remainingTimeMs = MAX_PERIODIC_TASK_STOP_TIME_MILLIS;
+      LOGGER.info("Task: {} is running, wait for at most {}ms for it to finish", _taskName, remainingTimeMs);
+      while (_running && remainingTimeMs > 0L) {
+        long sleepTimeMs = Long.max(remainingTimeMs, 1000L);
```
I'm not quite sure the logic here. The variable `remainingTimeMs` is initially set to 30_000L. Then `sleepTimeMs` is set to 30_000L, since it's assigned from the larger number between 30_000L and 1_000L. Then `remainingTimeMs` is set to 0L. In this case, the thread will be slept only once. Could it be `min()` instead of `max()`?

Jackie-Jiang replied:
Good catch! It should be min()

jackjlli commented on pinot-controller/src/test/java/org/apache/pinot/controller/helix/core/retention/RetentionManagerTest.java:
```diff
@@ -92,7 +92,7 @@ private void testDifferentTimeUnits(long pastTimeStamp, TimeUnit timeUnit, long
     conf.setDeletedSegmentsRetentionInDays(0);
     RetentionManager retentionManager = new RetentionManager(pinotHelixResourceManager, conf, controllerMetrics);
-    retentionManager.init();
+    retentionManager.start();
```
I'm thinking whether the method `stop` be added at the end but it doesn't seem matter.

Jackie-Jiang replied:
I have the same feeling, but I'm just lazy lol

jackjlli commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/relocation/RealtimeSegmentRelocator.java:
```diff
@@ -62,32 +61,13 @@ public RealtimeSegmentRelocator(PinotHelixResourceManager pinotHelixResourceMana
   protected void processTable(String tableNameWithType) {
-    CommonConstants.Helix.TableType tableType = TableNameBuilder.getTableTypeFromTableName(tableNameWithType);
-    if (tableType == CommonConstants.Helix.TableType.REALTIME) {
+    if (TableNameBuilder.REALTIME.tableHasTypeSuffix(tableNameWithType)) {
```
👍

sunithabeeram commented:
@Jackie-Jiang what race condition was addressed in this PR and how?
Was any manual testing done to make sure leader and non-leader controllers behave as expected?

mcvsubbu commented:
@Jackie-Jiang could you also please check with @npawar ? I think we had similar questions (should there be a stop/start, etc.) in her PR and had made some choices because of the way things were at the time. Thanks.

npawar commented on pinot-core/src/main/java/org/apache/pinot/core/periodictask/BasePeriodicTask.java:
```diff
@@ -18,20 +18,39 @@
 public abstract class BasePeriodicTask implements PeriodicTask {
+  private static final Logger LOGGER = LoggerFactory.getLogger(BasePeriodicTask.class);
+
+  // Wait for at most 5 minutes while calling stop() for task to terminate
```
comment says 5 minutes, value says 30s?

Jackie-Jiang replied:
Missed a zero. Change comment to 30s. Thanks

npawar commented on pinot-core/src/main/java/org/apache/pinot/core/periodictask/BasePeriodicTask.java:
```diff
@@ -42,9 +61,126 @@ public long getInitialDelayInSeconds() {
+      try {
+        runTask();
+      } catch (Exception e) {
+        LOGGER.error("Caught exception while running task: {}", _taskName, e);
```
Seems like a good place to introduce a metric to count failures in PeriodicTasks, so that we know if a task is erroring out

Jackie-Jiang replied:
Added metrics for CONTROLLER_PERIODIC_TASK_RUN and CONTROLLER_PERIODIC_TASK_ERROR to track both number of runs and errors

npawar replied:
Not setting them here? Or planning to handle it later? If so, please add a TODO

npawar commented on pinot-core/src/main/java/org/apache/pinot/core/periodictask/BasePeriodicTask.java:
```diff
@@ -42,9 +61,126 @@ public long getInitialDelayInSeconds() {
+        }
+      }
+      long waitTimeMs = System.currentTimeMillis() - startTimeMs;
+      if (_running) {
```
the LOG in if and else look like they're switched

Jackie-Jiang replied:
Correct, good catch

npawar commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/periodictask/ControllerPeriodicTask.java:
```diff
@@ -31,152 +31,90 @@
 
   /**
-   * This method runs before processing all tables
+   * Can be override to provide context before processing the tables.
```
*can be overridden

Jackie-Jiang replied:
Fixed

npawar replied:
I think you missed it again :) Here and also in 2 other places in this file

npawar commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/SegmentStatusChecker.java:
```diff
@@ -69,58 +65,50 @@ public SegmentStatusChecker(PinotHelixResourceManager pinotHelixResourceManager,
-  }
-
-  @Override
-  protected void exceptionHandler(String tableNameWithType, Exception e) {
```
Why remove exceptionHandler method? It was consciously added so that we avoid such try catch in each ControllerPeriodicTask.

Jackie-Jiang replied:
After consideration, I still think this exception handling logic should be wrapped inside the processTable(). It might need some variables from the processTable(), which are not accessible with this callback handler.

npawar commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/minion/PinotTaskManager.java:
```diff
@@ -44,134 +43,126 @@
-  public void stopTask() {
-    LOGGER.info("Perform task cleanups.");
-    // Performs necessary cleanups for each task type.
+  private void nonLeaderCleanUp() {
```
Can some comments be added, either here or above cleanupTask explaining why non leader cleanup is needed in this task? Since these tasks run on controller leadership gain, it's confusing to see this method here

Jackie-Jiang replied:
Added comments in BasePeriodicTask

npawar commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/periodictask/ControllerPeriodicTask.java:
```diff
@@ -31,152 +31,90 @@
  * The base periodic task for pinot controller only. It uses <code>PinotHelixResourceManager</code> to determine
  * which table resources should be managed by this Pinot controller.
+ *
+ * @param <C> the context type
```
Nicely done with the Context

npawar commented on pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/periodictask/ControllerPeriodicTask.java:
```diff
@@ -31,152 +31,90 @@
+      try {
+        processTable(tableNameWithType, context);
+      } catch (Exception e) {
+        LOGGER.error("Caught exception while processing table: {} in task: {}", tableNamesWithType, _taskName, e);
```
why not keep exceptionHandler and call it here?

Jackie-Jiang replied:
Added back processTableExceptionHandler and pass in Context

Jackie-Jiang commented:
@sunithabeeram The race condition between start() and stop(), and is addressed by adding synchronization.
As for the leader and non-leader change, ControllerPeriodicTaskTest covers the test of start() and stop(), which are called in ControllerPeriodicTaskScheduler onBecomingLeader() and onBecomingNonLeader() (implementing LeadershipChangeSubscriber).

Jackie-Jiang commented:
@npawar Addressed the comments. Periodic task logic is covered by unit test: ControllerPeriodicTask and PeriodicTaskSchedulerTest. For each individual periodic task, it's out of the scope of this PR, and we can add tests later.
